### ISSUE TITLE
A reader step whose group is 'trigger' does not execute the first item in an iterator.

### ISSUE THREAD

**Issue by matthewhilton** [catalyst/moodle-tool_dataflows#663]
While working on #31 I came across a behaviour that I'm not sure is a bug or if i'm doing something I shouldn't.

Basically I have a `reader_step` that I want to appear as a trigger, so i've set the `get_group` function to return `trigger`. This makes it appear as a trigger in the UI which is what I want.

I then have an iterator in the step that reads some stuff from the database. But I found that when i set the group to `trigger` it only calls the execute function N - 1 times. For e.g. if returns 4 records, then the execute function is only called 3 times, the first item in the iterator is always skipped. If there is just 1 record, nothing is run. Note I checked if it was making an adhoc task for the first one, but it isn't.

If I change the group type to `reader`, they all execute as expected.

I think either get_group should:
1. Be more clear that its not just UI but also logic.
2. Refactor the logic that is relying on `get_group`. For e.g. if its expecting an iterator but checking if its group is a `reader` before getting the iterator, it should just check for the existence of an iterator instead (this might not be the case, just a random idea)

**Comment by matthewhilton** [comment]
https://github.com/catalyst/moodle-tool_dataflows/blob/05a7ed801bd2cb7674a3fd16530abdd70fb597dc/classes/local/execution/iterators/dataflow_iterator.php#L188-L192

The first item is not being called here. There issue is likely further up the chain.

**Comment by matthewhilton** [comment]
I've worked it out.

Here it checks if `should_pull_next` as a way to NOT increment the iterator on the first run.

https://github.com/catalyst/moodle-tool_dataflows/blob/05a7ed801bd2cb7674a3fd16530abdd70fb597dc/classes/local/execution/iterators/dataflow_iterator.php#L159-L170

However this function under the hood 

https://github.com/catalyst/moodle-tool_dataflows/blob/05a7ed801bd2cb7674a3fd16530abdd70fb597dc/classes/local/execution/iterators/dataflow_iterator.php#L132-L140

Is checking the grouptype is `!== readers`, which is why any non-reader steps are skipping the first iteration in in iterator.

**Comment by matthewhilton** [comment]
Confirmed that removing the steptype check in `should_pull_next` fixes the bug. However, I am unsure if this may lead to regressions elsewhere without understanding the reason why this check was there in the first place.

**Comment by keevan** [comment]
Internal comms but here for completness:

The pulling of the next record was how it was originally added. Given an example of 3 steps, first one being a reader. The 3rd step would then pull the iterator for the next() record, which will cause step 2 to pull it's iterator, and cause step 1 to call its next function. And because step 1 is the reader, it doesn't pull the underlying $input's next value, because it needs to pass through the first record.

So in a nutshell, it's calling next() to grab the iterator value after processing it for the step before it.

**Comment by matthewhilton** [comment]
I think then it might be a good idea to have a new function for the UI grouping that by default just returns `return get_group()` but can be overridden if necessary (such as this case)
Then refactor https://github.com/catalyst/moodle-tool_dataflows/blob/05a7ed801bd2cb7674a3fd16530abdd70fb597dc/step-chooser.php#L75-L87 to use this new function instead

**Comment by keevan** [comment]
I wonder if flow_triggers / reader_triggers need to be split out into its own thing. It's definitely not the current norm, but there might be reason to have it (e.g. you might want to process some queued up events in a single run instead of one run per matching event).

Overriding get_group and using it for 2 different things feels incorrect. I'd lean more towards reading code where it's clear that for readers, and "trigger + readers", it will not pull the next record for its own iterator by default.

The current suggestion also might cause some concerns / more complexity with nested readers and what could be done about that, but I think that's currently unresolved(?) as well.

**Comment by matthewhilton** [comment]
> you might want to process some queued up events in a single run instead of one run per matching event

In this case you could just add some logic to the iterator to instead return an array on each iteration instead. Although i'm not sure if later steps are able to handle this.

> Overriding get_group and using it for 2 different things feels incorrect. 

We wouldn't be overriding `get_group` at all, we would be overriding a function that is used purely for the UI e.g.
```
function get_ui_grouping() {
    return $this->get_group()
}
```

I think for the time being the event reader step https://github.com/catalyst/moodle-tool_dataflows/pull/664 will just live as a reader, since that is what it technically is.

**Comment by matthewhilton** [comment]
After I rebased my branch onto the latest version and fixed up some things, this issue seems to be resolved ? I'm not able to replicate it anymore.
